### ISSUE TITLE
dmypy check support -c flag

### ISSUE THREAD

**Issue by nathanpage-credo** [python/mypy#15615]
**Feature**

(A clear and concise description of your feature proposal.)

mypy supports a `-c` flag to pass in code to check by string. dmypy is missing this feature.

**Pitch**
Add in the same option for dmypy

```bash
dmypy check -c "a=2"
```

**Comment by kcornw** [comment]
Hello! Thank you for bringing up this feature request. It seems that `dmypy` is designed to pass flags using the `run` command. You can achieve the desired functionality by using the following command:

```
dmypy run -- -c "a=2"
```

This approach leverages `mypy`'s existing `-c` flag through dmypy, allowing to check code snippets without creating temporary files.

However, there seems to be a bug where `dmypy` caches the result of the first `-c` check and does not re-evaluate subsequent `-c` checks.

```
$ dmypy run -- -c "a=1"
Success: no issues found in 1 source file

$ dmypy run -- -c "a:str=1"
Success: no issues found in 1 source file
```

It seems that `dmypy` incorrectly reused the cache?

**Comment by kcornw** [comment]
Please see #18326 about the bug I mentioned above
